design/proposals/ambient-namespaces.md

```diff
@@ -0,0 +1,126 @@
+  nginx.parts.serverBlockConfigMap(namespace, appName),
+])
+```
+ 
+This example does not lend well to the ksonnet desired model where components are environment agnostic; meaning a component should not need information on which environment the resulting Kubernetes Objects will be deployed to. Yet, it is unavoidable that these component / prototype configurations may need to reference a namespace, server URI, or other metadata.
+
+## Goals
+
+* Encourage behavior for component and prototype manifests to be environment-agnostic. 
+  Example: It is encouraged to write `nginx.parts.service(env.namespace, appName)`, but not `local namespace = "foo"` or `if (env.namespace == "foo") ... else ...`.
+* Component and prototype manifests should be able to refer to environment metadata fields, such as the environment namespace in the form of `env.namespace`.
+* Environment fields should be evaluated at run-time depending on which environment the user is trying to deploy to. 
+  Example: If the user had a `prod` environment configured to the namespace `prod-ns`, running `ks apply prod` should evaluate `env.namespace` to the value `prod-ns`.
+* Produce a second version of existing prototypes without hard-coded namespace values.
+
+## Non Goals
+
+* Access to a specific environment's metadata. 
+  Example: There should not be access to an attribute like `env.prod.namespace`. The evaluation of `env.namespace` should happen on `apply`, `diff`, or other ks functionalities that operate on an environment.
+
+## Proposal
+
+### Fundamentals
+
+The change will behave similarly to how parameters operate in component and prototype manifests.
+
+Components created on the `generate` command will introduce the following line to the top of the component manifest [1]:
+
+`local env = std.extVar("__ksonnet/env");`
+
+Namespace, Server URI, and other environment metadata attributes can then be accessed using `env.namespace` and `env.server`, respectively.
+
+For example, the components generated from the nginx prototype in the Motivation section above may appear as:
+
+```jsonnet
+local env = std.extVar("__ksonnet/env");
+
+local k = import 'k.libsonnet';
+local nginx = import '../nginx.libsonnet';
+
+local appName = "nginx-app";
+
+k.core.v1.list.new([
+  nginx.parts.deployment.withServerBlock(env.namespace, appName),
+  nginx.parts.service(env.namespace, appName),
+  nginx.parts.serverBlockConfigMap(env.namespace, appName),
+])
+```
+
+The corresponding prototype manifest would appear as:
+
+```jsonnet
+local k = import 'k.libsonnet';
+local nginx = import '../nginx.libsonnet';
+
+local appName = "nginx-app";
+
+k.core.v1.list.new([
+  nginx.parts.deployment.withServerBlock(import 'env://namespace', appName),
+  nginx.parts.service(import 'env://namespace', appName),
+  nginx.parts.serverBlockConfigMap(import 'env://namespace', appName),
+])
+```
+
+ksonnet would translate `import 'env://namespace'` prototype snippets to `env.namespace` during the `generate` command.
```

Comment:
Cool. Just so I understand: the presence of `env://namespace` in a prototype implies an optional flag, `--namespace`, and if that flag is not set, it points at `env.namespace`. Is that right? If so, that makes sense, and we probably want to say that explicitly.

(BTW, I suggest adding the flag only if `env://namespace` is present, because having an optional parameter that does nothing seems odd.)

Reply:
Hmm, I think my initial intention was to disallow users from being able to set environment metadata through the `generate` command altogether. Ex: Disallowing things like `ks generate some-prototype --namespace=foo`. That way it forces the namespace to be determined during `apply` and what not.

I chose the syntax `import 'env://namespace'` to follow the current custom conventions used by `import 'param://foo'` but there is no reason it can't directly be `env.namespace` (although this might not compile? I'm not sure).

What are your thoughts?